core/src/main/java/org/apache/calcite/rel/rel2sql/SqlImplementor.java

```diff
@@ -1437,6 +1437,8 @@ public static SqlNode toSql(RexLiteral literal) {
       return SqlLiteral.createTimestamp(typeName,
           castNonNull(literal.getValueAs(TimestampString.class)),
           literal.getType().getPrecision(), POS);
+    case BINARY:
+      return SqlLiteral.createBinaryString(requireNonNull(literal.getValueAs(byte[].class)), POS);
```

Comment:
NIT: why not use `castNonNull` to keep it the same with other types?

Reply:
I'm wondering if `select cast(null as binary)` would work for current implementation. It's true that all other types does not consider `null`, they may suffer from the same problem.

Reply:
> I'm wondering if `select cast(null as binary)` would work for current implementation. It's true that all other types does not consider `null`, they may suffer from the same problem.

@mihaibudiu What's your thought on this comment?

Reply:
I have added a test for this. It works.

Reply:
Seems that the `SqlTypeName` of a `RexLiteral` is always `NULL`, no matter what the `RelDataType` is.

Reply:
I think it's much worse, the typeName of RexLiteral is not null, it is used in a very ad-hoc way. It almost never matches the actual type.

The field's Javadoc discusses this:
```
// TODO jvs 26-May-2006:  Use SqlTypeFamily instead; it exists
  // for exactly this purpose (to avoid the confusion which results
  // from overloading SqlTypeName).
  /**
   * An indication of the broad type of this literal -- even if its type isn't
   * a SQL type. Sometimes this will be different than the SQL type; for
   * example, all exact numbers, including integers have typeName
   * {@link SqlTypeName#DECIMAL}. See {@link #valueMatchesType} for the
   * definitive story.
   */
```

So the typeName is used instead of a TypeFamily. I personally think this field should be completely removed, but there are many uses in the code that would have to be audited. It should have been deprecated in 2006. Moreover, removing it may break third-party software that depends on Calcite that is not tested as part of the Calcite testing process. The `type` is the right field to use.

The TypeFamily itself has a very unclear meaning: https://issues.apache.org/jira/browse/CALCITE-5986

I have tried to clean up the TypeFamily in #3411, but I couldn't do it systematically, it's now also used itself in many places.